hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/lock/OzoneManagerLock.java

```diff
@@ -0,0 +1,312 @@
+          throw ex;
+        }
+      } else if (compare > 0) {
+        // If this locking fails, we throw exception to user.
+        manager.lock(oldUserResource);
+        try {
+          manager.lock(newUserResource);
+        } catch (Exception ex) {
+          // We got an exception acquiring 2nd user lock. Release already
+          // acquired user lock, and throw exception to the user.
+          manager.unlock(oldUserResource);
+          throw ex;
+        }
+      } else {
+        // both users are equal.
+        manager.lock(oldUserResource);
+      }
+      lockSet.set(resource.setLock(lockSet.get()));
+    }
+  }
+
+  /**
+   * Acquire lock on multiple users.
+   * @param oldUserResource
+   * @param newUserResource
+   */
+  public void releaseMultiUserLock(String oldUserResource,
+      String newUserResource) {
+    Resource resource = Resource.USER;
+    int compare = newUserResource.compareTo(oldUserResource);
+    if (compare < 0) {
+      manager.unlock(newUserResource);
+      manager.unlock(oldUserResource);
+    } else if (compare > 0) {
+      manager.unlock(oldUserResource);
+      manager.unlock(newUserResource);
+    } else {
+      // both users are equal.
+      manager.unlock(oldUserResource);
+    }
+    lockSet.set(resource.clearLock(lockSet.get()));
+  }
+
+
+  public void releaseLock(String resourceName, Resource resource) {
+
+    // TODO: Not checking release of higher order level lock happened while
+    // releasing lower order level lock, as for that we need counter for
+    // locks, as some locks support acquiring lock again.
+    manager.unlock(resourceName);
+    // clear lock
+    lockSet.set(resource.clearLock(lockSet.get()));
+
+  }
+
+  /**
+   * Resource defined in Ozone.
+   */
+  public enum Resource {
+    // For S3 Bucket need to allow only for S3, that should be means only 1.
+    S3_BUCKET((byte) 0, "S3_BUCKET"), // = 1
+
+    // For volume need to allow both s3 bucket and volume. 01 + 10 = 11 (3)
+    VOLUME((byte) 1, "VOLUME"), // = 2
+
+    // For bucket we need to allow both s3 bucket, volume and bucket. Which
+    // is equal to 100 + 010 + 001 = 111 = 4 + 2 + 1 = 7
+    BUCKET((byte) 2, "BUCKET"), // = 4
+
+    // For user we need to allow s3 bucket, volume, bucket and user lock.
+    // Which is 8  4 + 2 + 1 = 15
+    USER((byte) 3, "USER"), // 15
+
+    S3_SECRET((byte) 4, "S3_SECRET"), // 31
+    PREFIX((byte) 5, "PREFIX"); //63
+
+    // level of the resource
+    private byte lockLevel;
+
+    // This will tell the value, till which we can allow locking.
+    private short mask;
+
+    // This value will help during setLock, and also will tell whether we can
+    // re-acquire lock or not.
+    private short setMask;
+
+    // Name of the resource.
+    private String name;
+
+    Resource(byte pos, String name) {
+      this.lockLevel = pos;
+      for (int x = 0; x < lockLevel + 1; x++) {
+        this.mask += (short) Math.pow(2, x);
```

Comment:
Ignore this comment. But I have to write this :). Since all we want to do is set all bits to one up to that level, that is for 0, bit we need to set 2^0 = 1, for the first bit we want both 0 and 1 set, which is 2^0 + 2^1 = 3 , and for the second bit 2^0 + 2^1 + 2^2 = 4 + 2 + 1 = 7. 
Can't we achieve the same by Math.pow(2, x +1) -1, that is for 2^3 -1 = 7, That would allow us to eliminate the loop. I know this is a nano-optimization, so feel free to ignore it; if it does not make sense.

Reply:
Done